### ISSUE TITLE
"regression" profile dv_ntsc(_wide) with decklink consumer uses reversed fieldorder + fix?

### ISSUE THREAD

**Issue by walterav1984** [mltframework/mlt#192]
Since issue #172  leads to multiple new fixes, test-cases and the discovery of this regression "NTSC reversed fieldorder requirement", I create this new issue for ease of testing/fixing.

REGRESSION:
During the testing of NTSC material for the #172 bug, I noticed that sometimes reversed playback (pressing 'h' key during melt playback) or shifting the fieldorder was needed until a videoclip played back smooth (without stutter/judder) using profile dv_ntsc(_wide) on a decklink consumer with CRT NTSC monitor. This raised some awareness since I wrongly created a upperfield first DV codec video which played back correctly in the current broken melt builds but did not play correct on firewire DV to CRT monitor converter. The opposite a lowerfield first DV video stutters in these broken melt builds but play back correct with older melt build and firewire DV to analog S-video CRT monitor.   
Older melt version 6.0.0 from the ubuntu 16.04 repo didn't express this strange shifted field order requirement for NTSC clips, the clips playback correct without stutter/judder. Testing multiple melt and interlaced NTSC samples, not only created by my self I can say that the current implementation of NTSC field handling is broken and introduced by the following commit 41519af back in October 2016 when introducing sliced3 processing.

FIX:
Restoring `if ( !progressive && m_displayMode->GetFieldDominance() == bmdUpperFieldFirst )` under 
"// Normal non-keyer playout - needs byte swapping" by replacing `if ( !progressive )` suggested by @max-verem fixes the correct field order handling back to its old state. This former code fix is still valid for mlt github master (tested yesterday). An alternative fix was also introduced by @max-verem  in a49c4ad and has also been tested correctly.

INCORRECTFIX?:
As a other mistake in converting PAL to NTSC footage, I made footage that was visually correctly upperfield first interlaced but with a progressive flag. Therefor it did not playback correctly in current melt builds and missed upsampling from 4:2:0 to 4:2:2 in @max-verem his version, but by forcing interlaced `+ frame->interlaced_frame = 1` detection that footage was also displaying correctly. This workaround may have unwanted side affects.

CLIPS
Try any of the NTSC interlaced colorbars (with moving square) from the following link.
http://www.gvgdevelopers.com/concrete/products/k2/test_clips/
Or convert clean.dv PAL example to NTSC lowerfield first example as test footage, see #172 for more examples:
`ffmpeg -r 29.97 -i clean.dv -filter:v "scale=interl=1,crop=in_w:480:0:48,fieldorder=bff,setdar=4:3" -acodec copy -target ntsc-dv cleanntscb.dv`

TESTING:
To check if PAL/NTSC interlacing field order is correct and limit the processing on the footage which can be done by third party software like melt(adobe/fcp/avid/edius) I use Decklink/Intensity "BlackmagicDesign Media Express" bundled software which only plays correctly interlaced/fieldorder footage in Linux using rawvideo uyvy422 pixfmt upperfield first directly to connected CRT monitors. For DV codec footage I use the program "test-dv" which is part of the iec61883lib example application DV (see test-dv https://github.com/walterav1984/dvswitch) and plays DV footage from firewire directly to ADVC S-video CRT monitor.

Software:
Ubuntu linux 16.04 amd64 
BMD Destkop video 10.8.4
melt version 6.0.0. (ubuntu repo) +all needed -dev packages from the ubuntu 16.04 repo to build melt

**Comment by ddennedy** [comment]
I do not understand where this commit with ID a49c4ad5a75ae7b9582bd556fb12c6b238cbc7ba lives! It is not in any branch, and it is not in my locally cloned repo.
Certainly, I see the problem with the current logic in the decklink consumer. The change mentioned in this commit is an improvement, but really the way to do this is the consumer sets a top_field_first property on itself to indicate the field order that it wants, and lets the auto-loaded fieldorder filter do the work instead of duplicating the code to the processing in itself.

**Comment by ddennedy** [comment]
```
--- src/modules/decklink/consumer_decklink.cpp
+++ src/modules/decklink/consumer_decklink.cpp
@@ -391,6 +391,7 @@ protected:
 			mlt_log_error( getConsumer(), "Profile is not compatible with decklink.\n" );
 			return false;
 		}
+		mlt_properties_set_int( properties, "top_field_first", m_displayMode->GetFieldDominance() == bmdUpperFieldFirst );
 
 		// Set the keyer
 		if ( m_deckLinkKeyer && ( m_isKeyer = mlt_properties_get_int( properties, "keyer" ) ) )
@@ -526,8 +527,6 @@ protected:
 
 			if ( buffer )
 			{
-				int progressive = mlt_properties_get_int( MLT_FRAME_PROPERTIES( frame ), "progressive" );
-
 				// NTSC SDI is always 486 lines
 				if ( m_height == 486 && height == 480 )
 				{
@@ -548,20 +547,6 @@ protected:
 					unsigned char *arg[3] = { image, buffer };
 					ssize_t size = stride * height;
 
-					// convert lower field first to top field first
-					if ( !progressive )
-					{
-						// Make the first line black for field order correction.
-						for ( int i = 0; i < m_width; i++ )
-						{
-							buffer[ i * 2 + 0 ] = 128;
-							buffer[ i * 2 + 1 ] = 16;
-						}
-
-						arg[1] += stride;
-						size -= stride;
-					}
-
 					// Normal non-keyer playout - needs byte swapping
 					if ( !m_sliced_swab )
 						swab2( arg[0], arg[1], size );
@@ -578,17 +563,6 @@ protected:
 					uint32_t* s = (uint32_t*) image;
 					uint32_t* d = (uint32_t*) buffer;
 
-					if ( !progressive && m_displayMode->GetFieldDominance() == bmdUpperFieldFirst )
-					{
-						// Make the first line black for field order correction.
-						memset( buffer, 0, stride );
-
-						// Correct field order
-						height--;
-						y--;
-						d += m_width;
-					}
-
 					// Need to relocate alpha channel RGBA => ARGB
 					while ( --y )
 					{
```

**Comment by max-verem** [comment]
> As a other mistake in converting PAL to NTSC footage, I made footage that was visually correctly upperfield first interlaced but with a progressive flag. Therefor it did not playback correctly in current melt builds and missed upsampling from 4:2:0 to 4:2:2 in @max-verem his version, but by forcing interlaced + frame->interlaced_frame = 1 detection that footage was also displaying correctly. This workaround may have unwanted side affects.

this relates to 4:2:0 conversion of file that previously during encoding seems lost interlacing flag - it hack rather just check what was wrong with method, but this hack make a clear - file is wrong (actually not correctly set interlaced flag)


> I do not understand where this commit with ID a49c4ad lives! It is not in any branch, and it is not in my locally cloned repo.

i postponed this, because i am trying to understand why it worked properly before this changes. decklink_consumer does not set *progressive* and *top_field_first* for appropriate mode. I see you patch set only one. On other hand i am warring about *transition_composite* properly, because i suspect it hard-coded (assuming) for lower field first mode:
```
	// Assuming lower field first
	// Special care is taken to make sure the b_frame is aligned to the correct field.
	// field 0 = lower field and y should be odd (y is 0-based).
	// field 1 = upper field and y should be even.
``` 

checking *top_field_first* frame properties to shift or not seems not a correct - correct would be to provide properly set *progressive* and *top_field_first* consumer properties....

**Comment by ddennedy** [comment]
>  i am trying to understand why it worked properly before this changes. 

The decklink consumer was using the mlt_consumer default of lower field first, the fieldorder filter normalized inputs to that, and the decklink consumer converted to upper field first if the mode required it.

Leave the composite transition out of this. That just adds confusion and other bugs to the mix. It's code predates the introduction of the fieldorder filter.

**Comment by walterav1984** [comment]
Don't know which code is best/correct, but I can only verify the practical and visual aspects of it. @ddennedy his code is also been working/visual correct for the NTSC interlaced part.

`--- src/modules/decklink/consumer_decklink.cpp
+++ src/modules/decklink/consumer_decklink.cpp
@@ -391,6 +391,7 @@ protected:
 			mlt_log_error( getConsumer(), "Profile is not compatible with decklink.\n" );
 			return false;
 		}
+		mlt_properties_set_int( properties, "top_field_first", m_displayMode->GetFieldDominance() == bmdUpperFieldFirst );
 
 		// Set the keyer
 		if ( m_deckLinkKeyer && ( m_isKeyer = mlt_properties_get_int( properties, "keyer" ) ) )
@@ -526,8 +527,6 @@ protected:
 
 			if ( buffer )
 			{
-				int progressive = mlt_properties_get_int( MLT_FRAME_PROPERTIES( frame ), "progressive" );
-
 				// NTSC SDI is always 486 lines
 				if ( m_height == 486 && height == 480 )
 				{
@@ -548,20 +547,6 @@ protected:
 					unsigned char *arg[3] = { image, buffer };
 					ssize_t size = stride * height;
 
-					// convert lower field first to top field first
-					if ( !progressive )
-					{
-						// Make the first line black for field order correction.
-						for ( int i = 0; i < m_width; i++ )
-						{
-							buffer[ i * 2 + 0 ] = 128;
-							buffer[ i * 2 + 1 ] = 16;
-						}
-
-						arg[1] += stride;
-						size -= stride;
-					}
-
 					// Normal non-keyer playout - needs byte swapping
 					if ( !m_sliced_swab )
 						swab2( arg[0], arg[1], size );
@@ -578,17 +563,6 @@ protected:
 					uint32_t* s = (uint32_t*) image;
 					uint32_t* d = (uint32_t*) buffer;
 
-					if ( !progressive && m_displayMode->GetFieldDominance() == bmdUpperFieldFirst )
-					{
-						// Make the first line black for field order correction.
-						memset( buffer, 0, stride );
-
-						// Correct field order
-						height--;
-						y--;
-						d += m_width;
-					}
-
 					// Need to relocate alpha channel RGBA => ARGB
 					while ( --y )
 					{
`

**Comment by max-verem** [comment]
@ddennedy please apply your patch

**Comment by walterav1984** [comment]
NTSC field order reverse issue fixed and tested successful in commit c954f1d and b01523a.
